### PR TITLE
Vorticity Auxiliary Target: Explicit Wake Structure Learning

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The aft-foil pressure distribution in tandem flow is largely determined by the fore-foil's **wake vorticity** reaching the aft-foil. Currently the model must learn this wake interaction implicitly through the pressure and velocity targets. By adding an auxiliary vorticity prediction head, we force the backbone to explicitly represent the wake structure — which directly conditions p_tan.

**Physical motivation:** Vorticity ω = ∂Uy/∂x - ∂Ux/∂y is zero everywhere in inviscid flow and non-zero only in the boundary layer and wake. The fore-foil wake is the critical feature for tandem pressure prediction. Predicting ω forces the model to localize this structure explicitly.

**Literature:** FINE (arXiv:2505.15329) showed that vorticity-streamfunction decomposition provides physically interpretable intermediate representations that improve OOD flow prediction. The Helmholtz decomposition separates rotational (vortical) and irrotational flow components.

### Step 1: Pre-compute vorticity targets using KNN finite differences

The vorticity ω = ∂Uy/∂x - ∂Ux/∂y can be computed from the existing velocity targets (Ux, Uy) using K-nearest-neighbor gradient estimation on the unstructured mesh.

**Pre-computation (add to data loading or as offline preprocessing):**

```python
from scipy.spatial import cKDTree
import numpy as np

def compute_vorticity(coords, Ux, Uy, K=8):
    """
    Compute vorticity at each node via KNN least-squares gradient.
    coords: (N, 2) node positions
    Ux, Uy: (N,) velocity components
    Returns: omega (N,) vorticity field
    """
    tree = cKDTree(coords)
    _, indices = tree.query(coords, k=K+1)
    indices = indices[:, 1:]  # exclude self, shape (N, K)

    N = coords.shape[0]
    omega = np.zeros(N)

    for i in range(N):
        nbr = indices[i]
        dx = coords[nbr, 0] - coords[i, 0]
        dy = coords[nbr, 1] - coords[i, 1]
        dUx = Ux[nbr] - Ux[i]
        dUy = Uy[nbr] - Uy[i]

        # Least-squares: [dx dy] @ [grad_x, grad_y]^T = dU
        A = np.column_stack([dx, dy])
        ATA = A.T @ A + 1e-10 * np.eye(2)

        # ∂Uy/∂x and ∂Ux/∂y
        grad_Uy = np.linalg.solve(ATA, A.T @ dUy)  # [∂Uy/∂x, ∂Uy/∂y]
        grad_Ux = np.linalg.solve(ATA, A.T @ dUx)  # [∂Ux/∂x, ∂Ux/∂y]

        omega[i] = grad_Uy[0] - grad_Ux[1]  # ∂Uy/∂x - ∂Ux/∂y

    return omega
```

**Vectorized version (MUCH faster, use this):**

```python
def compute_vorticity_vectorized(coords, Ux, Uy, K=8):
    """Vectorized vorticity computation — processes all nodes at once."""
    tree = cKDTree(coords)
    _, indices = tree.query(coords, k=K+1)
    indices = indices[:, 1:]  # (N, K)

    N = coords.shape[0]
    nbr_coords = coords[indices]  # (N, K, 2)
    delta = nbr_coords - coords[:, None, :]  # (N, K, 2)

    dUx = Ux[indices] - Ux[:, None]  # (N, K)
    dUy = Uy[indices] - Uy[:, None]

    # Batch least-squares via normal equations
    ATA = np.einsum('nki,nkj->nij', delta, delta)  # (N, 2, 2)
    ATA[:, 0, 0] += 1e-10
    ATA[:, 1, 1] += 1e-10

    ATdUx = np.einsum('nki,nk->ni', delta, dUx)  # (N, 2)
    ATdUy = np.einsum('nki,nk->ni', delta, dUy)

    # Solve for gradients
    grad_Ux = np.linalg.solve(ATA, ATdUx)  # (N, 2): [∂Ux/∂x, ∂Ux/∂y]
    grad_Uy = np.linalg.solve(ATA, ATdUy)  # (N, 2): [∂Uy/∂x, ∂Uy/∂y]

    omega = grad_Uy[:, 0] - grad_Ux[:, 1]  # ∂Uy/∂x - ∂Ux/∂y
    return omega
```

**Caching strategy:** Pre-compute vorticity for ALL training and validation samples and cache as `.npy` files (keyed by sample index). This takes ~5-10 minutes for the full dataset and only needs to run once.

**Alternative: on-the-fly computation.** If pre-caching is difficult, compute vorticity on-the-fly in the data loader. The KNN + gradient computation is ~5ms per sample on CPU, adding ~10% to data loading time.

### Step 2: Add vorticity auxiliary head

Add a small MLP that predicts vorticity from the backbone hidden representation:

```python
class VorticityHead(nn.Module):
    def __init__(self, hidden_dim=192, out_dim=1, layers=2):
        super().__init__()
        self.mlp = nn.Sequential(
            nn.Linear(hidden_dim, hidden_dim // 2),
            nn.GELU(),
            nn.Linear(hidden_dim // 2, out_dim)
        )

    def forward(self, h):
        # h: (B, N, D) backbone hidden representation
        return self.mlp(h).squeeze(-1)  # (B, N)
```

This adds ~18K parameters. Apply it to the backbone's output (same hidden representation that feeds into the pressure/velocity heads).

### Step 3: Auxiliary loss

```python
# In training loop:
omega_pred = vorticity_head(backbone_hidden)
omega_gt = batch['vorticity']  # pre-computed target

# Clip extreme vorticity values to reduce noise from FD on mesh boundaries
omega_gt_clipped = omega_gt.clamp(-vort_clip, vort_clip)

L_vort = lambda_v * F.l1_loss(omega_pred, omega_gt_clipped)
total_loss = standard_loss + L_vort
```

**Vorticity clipping:** Finite-difference vorticity on unstructured meshes can have extreme values at mesh boundaries and near sharp geometry features (leading edge, trailing edge). Clip to ±3σ of the training distribution to prevent these outliers from dominating the loss. Compute the clipping threshold from a sample of 100 training cases.

### Step 4: Flags and experiments

Flags:
- `--aux_vorticity`: Enable vorticity auxiliary target
- `--aux_vorticity_weight <float>`: Loss weight (default 0.1)
- `--vorticity_clip <float>`: Clipping threshold for vorticity targets

Run **2 seeds** per config:

```bash
# Config A: weight=0.05 (gentle)
cd cfd_tandemfoil && python train.py --agent frieren --seed 42 \
  --wandb_name "frieren/vort-aux-w0.05-s42" \
  --wandb_group "frieren/vorticity-aux" \
  --aux_vorticity --aux_vorticity_weight 0.05 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias

# Config B: weight=0.1 (standard)
# Same flags but --aux_vorticity_weight 0.1
# Seed 73: change --seed 73
```

### Key implementation notes

1. **Velocity channel indices:** In the model output, identify which channels are Ux and Uy. The vorticity computation uses the GROUND TRUTH velocity from the dataset, not the model's predictions (for the target). The model PREDICTS vorticity from the backbone hidden state.

2. **Coordinate extraction:** Node positions are in `x[:, :, :2]` (first 2 input channels). Use these for KNN computation.

3. **torch.compile:** The VorticityHead is a simple MLP — fully compile-compatible. The vorticity pre-computation runs in numpy on CPU.

4. **Gradient flow:** The vorticity loss gradient flows into the shared backbone but NOT into the pressure/velocity heads. This is the desired behavior — the auxiliary loss shapes the backbone representation without directly affecting the primary prediction heads.

5. **PCGrad interaction:** The vorticity loss is an auxiliary loss. It should be included in the same backward pass as the primary loss (not as a separate PCGrad group). PCGrad operates on the 2-way single-foil vs tandem split, which is unchanged.

6. **VRAM:** The VorticityHead adds ~18K params (~72KB). Negligible VRAM impact.

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| p_in | **13.05** | < 13.05 |
| p_oodc | **7.70** | < 7.70 |
| **p_tan** | **28.60** | **< 28.60** |
| p_re | **6.55** | < 6.55 |

**Baseline W&B runs:** d7l91p0x (s42, p_tan=28.9), j9btfx09 (s73, p_tan=28.3)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent frieren --wandb_name "frieren/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2183 because Closing: No progress after 8+ hours — zero W&B runs, no implementation commits. Vorticity KNN finite differences on unstructured mesh is complex and will be deferred to a future round. Frieren is being reassigned to a more tractable experiment..

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Config A: aux_vorticity_weight=0.05

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| p_in | 13.5 | 13.0 | **13.25** | 13.05 | +1.5% |
| p_tan | 29.3 | 29.1 | **29.20** | 28.60 | **+2.1%** |
| p_oodc | 7.9 | 7.9 | **7.90** | 7.70 | +2.6% |
| p_re | 6.5 | 6.5 | **6.50** | 6.55 | -0.8% |

### Config B: aux_vorticity_weight=0.1

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| p_in | 13.4 | 13.5 | **13.45** | 13.05 | +3.1% |
| p_tan | 29.4 | 29.3 | **29.35** | 28.60 | **+2.6%** |
| p_oodc | 7.7 | 7.7 | **7.70** | 7.70 | 0.0% |
| p_re | 6.5 | 6.2 | **6.35** | 6.55 | -3.1% |

### Implementation details
- **Vorticity pre-computation:** KNN (K=8) least-squares gradient estimation on unstructured mesh, vectorized via numpy batch solve. Pre-computed for all 1322 training samples in ~5.5 min; stored in dataset wrapper with custom pad_collate.
- **Normalization:** Vorticity normalized to zero-mean/unit-std (mean=417.4, std=22889), then clipped at ±3σ.
- **VorticityHead:** 18,625 params (Linear(192→96) → GELU → Linear(96→1)), zero-initialized output.
- **Loss integration:** L1 loss on masked nodes, added as shared auxiliary to both PCGrad groups at half weight each.
- **Performance:** 72-80s/epoch (identical to baseline), 46-47GB VRAM (negligible overhead).
- **Training:** All runs hit 180-min wall-clock timeout at epoch 149/500. Results represent incomplete training convergence.

### W&B runs
- w=0.05, s42: `vw3jpr15` — p_tan=29.3
- w=0.05, s73: `lo2dayem` — p_tan=29.1
- w=0.1, s42: `avivx0eh` — p_tan=29.4
- w=0.1, s73: `9afvei72` — p_tan=29.3

### Commands
```bash
# Config A (w=0.05)
cd cfd_tandemfoil && python train.py --agent frieren --seed 42 \
  --wandb_name "frieren/vort-aux-w0.05-s42" --wandb_group "frieren/vorticity-aux" \
  --aux_vorticity --aux_vorticity_weight 0.05 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias
# Config B: same but --aux_vorticity_weight 0.1
```

### Peak memory
46.1-47.1 GB (baseline: ~45-46 GB; +1 GB from vorticity cache in RAM, not GPU)

### Analysis

**Negative result.** Both vorticity auxiliary weights hurt p_tan by +2-3% while having mixed effects on other metrics:
- **p_re improved** slightly with w=0.1 (6.35 vs 6.55 baseline, -3.1%), suggesting the vorticity signal helps OOD Reynolds generalization.
- **p_tan degraded** consistently across all 4 runs (29.1-29.4 vs 28.6 baseline), indicating the auxiliary loss competes with rather than complements the primary tandem pressure objective.
- **p_oodc was neutral** at w=0.1 (7.70 = baseline) but slightly worse at w=0.05.

**Why it likely failed:**
1. **Gradient competition in the backbone.** The vorticity loss gradient through the shared backbone competes with the primary pressure/velocity gradients. Even at low weight (0.05), it diverts backbone capacity toward representing vorticity at the expense of the pressure-optimal representation already learned by the well-tuned baseline.
2. **Vorticity is already implicitly captured.** The baseline model's backbone already learns wake structure implicitly — the velocity targets themselves encode the vorticity information (ω = curl(v)). Adding an explicit vorticity head is redundant rather than complementary.
3. **Noisy FD targets on unstructured meshes.** KNN-based vorticity estimation is inherently noisy, especially at mesh boundaries and resolution transitions. Even with ±3σ clipping, this noise may inject harmful gradients into the backbone.
4. **Incomplete training (149/500 epochs).** All runs were cut short by the 180-min timeout. If the vorticity auxiliary helped backbone convergence speed, this wouldn't show up in truncated runs. However, at epoch 149 the baseline is also still converging, so the relative comparison is still informative.

### Suggested follow-ups
1. **Detached vorticity head:** Detach the hidden representation before feeding to VorticityHead (no gradient flow back to backbone). This would test whether the vorticity prediction is useful as a diagnostic without disturbing the backbone.
2. **Vorticity as an input feature:** Pre-compute vorticity and add it as an input feature to the model rather than as an auxiliary target. This gives the model explicit vorticity information without the gradient competition issue.
3. **Tandem-only vorticity loss:** Apply the auxiliary loss only to tandem samples where wake structure matters most, avoiding interference with single-foil training.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2183 because ADVISOR: Thank you for the thorough analysis and for implementing the KNN vorticity estimation despite the complexity — the implementation quality is excellent.

**Decision: Closing as dead end.**

Results vs current baseline (PR #2184, p_tan=28.502):
- Config A (w=0.05): p_tan=29.20 (+2.5%), p_in=13.25 (+0.3%), p_oodc=7.90 (+1.1%), p_re=6.50 (+0.7%) — all metrics worse
- Config B (w=0.1): p_tan=29.35 (+3.0%) — primary metric worse despite p_oodc/p_re improvements

Your analysis is correct on all three failure modes:
1. **Gradient competition in the shared backbone** is the primary issue — the vorticity gradient diverts capacity from the already well-optimized pressure representation
2. **Vorticity is implicitly encoded** in the velocity targets (ω = curl(v)) — the baseline already sees this signal indirectly
3. **Noisy FD targets on unstructured meshes** inject harmful gradients even at low weight

The incomplete training (149/500 epochs due to wall-clock timeout) doesn't change the conclusion — the relative degradation at epoch 149 is consistent and well-understood.

**Most promising follow-up (for a future round):** Your suggestion to use vorticity as an *input feature* (pre-computed, added to node features) rather than as an auxiliary loss target is architecturally cleaner — it gives the model wake information without gradient interference. We'll keep this in the hypothesis queue.

You are currently assigned to PR #2199 (Spectral conditioning of attention). Continue with that experiment..
